src/vs/workbench/services/files/electron-browser/fileService.ts

```diff
@@ -43,7 +43,7 @@ export class FileService implements IFileService {
 	private activeOutOfWorkspaceWatchers: ResourceMap<uri>;
 
 	protected _onFileChanges: Emitter<FileChangesEvent>;
-	private _onAfterOperation: Emitter<FileOperationEvent>;
+	protected _onAfterOperation: Emitter<FileOperationEvent>;
```

Comment:
@bpasero Why are there *afterOperation*s for some things (create, delete) but not for all (change) and why do we have this at all and why not rely on file events?

Reply:
@jrieken file events can come with a delay and if we would update the explorer based on these events, any operation would appear dead slow. So we have the "local" events that fire immediately after a file operation succeeded to quickly update the explorer. We do not need these events for changes within files, only structural changes that have an impact on the explorer.

Reply:
> So we have the "local" events that fire immediately after a file operation succeeded to quickly update the explorer. 

Why isn't the promise it returns that "local" event?

Reply:
Some clients trigger file service methods (like the actions) from outside the explorer, so the event is needed to decouple the updating of the explorer. But maybe I did not get the suggestion from you correctly?

Reply:
> Some clients trigger file service methods (like the actions) from outside the explore

Good point, didn't think about that. I am confused by two things: (1) these events often already contains the (target) stat e.g for create/move/copy and (2) if I don't send those event but normal file events nothing updates... So, yes the notion of an *onWillModify*-event makes sense but only if it doesn't contain the actual stat yet

Reply:
Well, typically after a file operation, the resource is resolved again (= stat) so it can be added to the event. The stat is needed for the explorer to update properly, e.g. when a new file is added the stat can just be added to the model without the need of resolving the parent folder again.

If you are sending file events and the explorer is not updating properly, please check with @isidorn. I would assume that ADD and DELETE events of resources that are visible in the explorer result in the explorer refreshing. Anything else seems like a bug to me that would also cause normal external file events to result in a stale explorer.

Reply:
> Well, typically after a file operation, the resource is resolved again (= stat) so it can be added to the event. 

Yeah, but *after* the operation I should have gotten the file event already? This is to only take slow file event into account but not slow fs-operations

Reply:
@jrieken yes, the file events always come later from my experience. There is lots of layers of buffering and delaying up until the event hits us. I never found the file event to come first.

Reply:
Assumptions might change given things are now really remote, e.g. creating and stat'ing a file on a ftp server.

Reply:
Yes, and the explorer needs to always be fit to handle either local or remote events in any order independent from the backend. The local event is just a hint to refresh faster if possible, but depending on the backend that might not be true.